src/trace_ng/log.zig

```diff
@@ -0,0 +1,1048 @@
+            self.log(.warn, message);
+        }
+
+        pub fn warnf(self: *Self, comptime fmt: []const u8, args: anytype) void {
+            self.logf(.warn, fmt, args);
+        }
+
+        pub fn warnWithFields(self: *Self, message: []const u8, fields: anytype) void {
+            self.logWithFields(.warn, message, fields);
+        }
+
+        pub fn warnfWithFields(self: *Self, comptime fmt: []const u8, args: anytype, fields: anytype) void {
+            self.logfWithFields(.warn, fmt, args, fields);
+        }
+
+        pub fn info(self: *Self, message: []const u8) void {
+            self.log(.info, message);
+        }
+
+        pub fn infof(self: *Self, comptime fmt: []const u8, args: anytype) void {
+            self.logf(.info, fmt, args);
+        }
+
+        pub fn infoWithFields(self: *Self, message: []const u8, fields: anytype) void {
+            self.logWithFields(.info, message, fields);
+        }
+
+        pub fn infofWithFields(self: *Self, comptime fmt: []const u8, args: anytype, fields: anytype) void {
+            self.logfWithFields(.info, fmt, args, fields);
+        }
+
+        pub fn debug(self: *Self, message: []const u8) void {
+            self.log(.debug, message);
+        }
+
+        pub fn debugf(self: *Self, comptime fmt: []const u8, args: anytype) void {
+            self.logf(.debug, fmt, args);
+        }
+
+        pub fn debugWithFields(self: *Self, message: []const u8, fields: anytype) void {
+            self.logWithFields(.debug, message, fields);
+        }
+
+        pub fn debugfWithFields(self: *Self, comptime fmt: []const u8, args: anytype, fields: anytype) void {
+            self.logfWithFields(.debug, fmt, args, fields);
+        }
+
+        pub fn log(self: *Self, level: Level, message: []const u8) void {
+            switch (self.*) {
+                .noop => {},
+                inline else => |*impl| impl.*.log(level, message),
+            }
+        }
+
+        pub fn logf(self: *Self, level: Level, comptime fmt: []const u8, args: anytype) void {
+            switch (self.*) {
+                .noop => {},
+                inline else => |impl| impl.logf(level, fmt, args),
+            }
+        }
+
+        pub fn logWithFields(self: *Self, level: Level, message: []const u8, fields: anytype) void {
+            switch (self.*) {
+                .noop => {},
+                inline else => |impl| impl.logWithFields(level, message, fields),
+            }
+        }
+
+        pub fn logfWithFields(self: *Self, level: Level, comptime fmt: []const u8, args: anytype, fields: anytype) void {
+            switch (self.*) {
+                .noop => {},
+                inline else => |impl| impl.logfWithFields(level, fmt, args, fields),
+            }
+        }
+    };
+}
+
+pub const Logger = ScoppedLogger(null);
+
+/// An instance of `ScopedLogger` that logs to the standard err.
+pub fn StandardErrLogger(comptime scope: ?[]const u8) type {
```

Comment:
You could make `scope` a parameter for the log functions instead of being a parameter for the logger implementation types. The logger interface would still need it, but the implementations of the interface do not need it.

Personally I feel that it simplifies the code because this can be defined as a normal struct instead of a generic struct. It cleans up an unnecessary layer of abstraction

Reply:
Not sure if/how that would work since the implementation contains logic that refers to the scope. For example:

```
    pub fn log(self: *Self, level: Level, message: []const u8) void {
        if (@intFromEnum(level) > @intFromEnum(self.max_level)) {
            // noop
            return;
        }

        const maybe_scope = if (scope) |s| s else null; // <---- The scope needed here
        const log_msg = logfmt.LogMsg{
            .level = level,
            .maybe_scope = maybe_scope,
            .maybe_msg = message,
            .maybe_fields = null,
            .maybe_fmt = null,
        };

        self.channel.send(log_msg) catch |err| {
            std.debug.print("Send msg through channel failed with err: {any}", .{err});
            return;
        };
    }
```

Reply:
Unless the method on the implementation be modified to take scope like this

```zig
pub fn log(self: *Self, comptime scope: ?[]const u8, level: Level, message: []const u8) void
```

and then modify the interface to supply the scope when calling the implementation.

Reply:
Done in [460cf27](https://github.com/Syndica/sig/pull/249/commits/460cf2779a39f19fe1a5ff26c15eccda90796ea1)